code/datums/material_container.dm

```diff
@@ -0,0 +1,251 @@
+/*
+	This datum should be used for handling mineral contents of machines and whatever else is supposed to hold minerals and make use of them.
+
+	Variables:
+		amount - raw amount of the mineral this container is holding, calculated by the defined value MINERAL_MATERIAL_AMOUNT=2000.
+		max_amount - max raw amount of mineral this container can hold.
+		sheet_type - type of the mineral sheet the container handles, used for output.
+		owner - object that this container is being used by, used for output.
+		MAX_STACK_SIZE - size of a stack of mineral sheets. Constant.
+*/
+
+/datum/material_container
+	var/total_amount = 0
+	var/max_amount
+	var/sheet_type
+	var/obj/owner
+	var/list/materials = list()
+	//MAX_STACK_SIZE = 50
+	//MINERAL_MATERIAL_AMOUNT = 2000
+
+/datum/material_container/New(obj/O, list/mat_list, max_amt = 0)
+	owner = O
+	max_amount = max(0, max_amt)
+
+	if(mat_list[MAT_METAL])
+		materials[MAT_METAL] = new /datum/material/metal()
+	if(mat_list[MAT_GLASS])
+		materials[MAT_GLASS] = new /datum/material/glass()
+	if(mat_list[MAT_SILVER])
+		materials[MAT_SILVER] = new /datum/material/silver()
+	if(mat_list[MAT_GOLD])
+		materials[MAT_GOLD] = new /datum/material/gold()
+	if(mat_list[MAT_DIAMOND])
+		materials[MAT_DIAMOND] = new /datum/material/diamond()
+	if(mat_list[MAT_URANIUM])
+		materials[MAT_URANIUM] = new /datum/material/uranium()
+	if(mat_list[MAT_PLASMA])
+		materials[MAT_PLASMA] = new /datum/material/plasma()
+	if(mat_list[MAT_BANANIUM])
+		materials[MAT_BANANIUM] = new /datum/material/bananium()
+
+//For inserting an amount of material
+/datum/material_container/proc/insert_amount(amt, material_type = null)
+	if(amt > 0 && has_space(amt))
+		var/total_amount_saved = total_amount
+		if(material_type)
+			for(var/datum/material/M in materials)
+				if(M.material_type == material_type)
+					M.amount += amt
+					total_amount += amt
+		else
+			for(var/datum/material/M in materials)
+				M.amount += amt
+				total_amount += amt
+		return (total_amount - total_amount_saved)
+	return 0
+
+/datum/material_container/proc/insert_stack(obj/item/stack/S, amt = 0)
+	if(!amt)
+		amt = S.amount
+	var/material_amt = get_item_material_amount(S)
+	amt = min(amt, round(((max_amount - total_amount) / material_amt)))
+	if(!amt)
+		return 0
+
+	for(var/i=0, i < amt, i++)
+		insert_materials(S)
```

Comment:
Better way would be adding optional argument to insert_materials proc that is 1 by default which is multiplier of the amount
So here you would just set the multiplier to amt